dbeaver/dbeaver#39997 Fix SQL execution error in IoTDB data source

LJW21-02 opened the pull request:
Primary changes in this PR:
- Fixed connection url parsing logic to resolve SQL execution error in IoTDB-tree connection
  <img width="839" height="146" alt="Screenshot 2025-12-25 184058" src="https://github.com/user-attachments/assets/0cc9fb59-90e4-495b-8398-a00928d4d572" />

  <img width="362" height="138" alt="Screenshot 2025-12-26 091639" src="https://github.com/user-attachments/assets/bebbd2d2-467d-46d8-8683-96aed9b19cc2" />

Other changes:
- Updated IoTDB JDBC version
- Removed unused privilege 
- Introduced auto-close with try-with-resources

LJW21-02 committed:
dbeaver/dbeaver#39997 Update JDBC version

LJW21-02 committed:
dbeaver/dbeaver#39997 Fix connection url parsing logic

LJW21-02 committed:
dbeaver/dbeaver#39997 Remove extra privilege

LJW21-02 committed:
dbeaver/dbeaver#39997 Introduce auto-close with try-with-resources

LJW21-02 commented:
@ShadelessFox Hi, this PR fixes an issue in the IoTDB plugin. Could you please help review it and let me know if any changes are needed? Thank you!

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/model/meta/IoTDBMetaModel.java:
```diff
@@ -59,21 +59,25 @@ public String getTableDDL(@NotNull DBRProgressMonitor monitor,
+            try (JDBCStatement stmt = session.createStatement()) {
+                try (JDBCResultSet rs = stmt.executeQuery(sql)) {
+                    while (rs.next()) {
+                        isAligned = (rs.getString("IsAligned")).equals("true");
```
This will assign `isAligned` to the value of the latest row if there is more than one row. If just one row is expected, consider using `if` instead of `while`.

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/model/meta/IoTDBTableMetaModel.java:
```diff
@@ -138,18 +138,22 @@ private String getTTL(DBRProgressMonitor monitor,
+            try (JDBCStatement stmt = session.createStatement()) {
+                try (JDBCResultSet rs = stmt.executeQuery(sql)) {
+                    while (rs.next()) {
+                        ttl = rs.getString("ttl(ms)");
```
Same issue as mentioned before.

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/model/meta/IoTDBTableMetaModel.java:
```diff
@@ -138,18 +138,22 @@ private String getTTL(DBRProgressMonitor monitor,
+                try (JDBCStatement stmt = session.createStatement()) {
+                    try (JDBCResultSet rs = stmt.executeQuery(sql)) {
+                        while (rs.next()) {
+                            ttl = rs.getString("TTL(ms)");
```
Same issue as mentioned before.

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/model/meta/IoTDBTableMetaModel.java:
```diff
@@ -171,26 +175,28 @@ private String getTableDDLInfoWithISPrivilege(JDBCSession session,
+                    }
+                    toAppend.append(",\n");
+                }
+                toAppend.setLength(toAppend.length() - 2);
```
This will truncate the opening paren if no columns were returned. Not sure if it's possible, though...

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/model/meta/IoTDBTableMetaModel.java:
```diff
@@ -171,26 +175,28 @@ private String getTableDDLInfoWithISPrivilege(JDBCSession session,
+                toAppend.setLength(toAppend.length() - 2);
+                String tableComment = ((DBSEntity) sourceObject).getDescription();
+                if (tableComment != null && !tableComment.isEmpty()) {
+                    toAppend.append("\n) COMMENT '").append(tableComment).append("' ");
```
Use `SQLUtils#quoteString` that does proper escaping instead of manually quoting the string.

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/model/meta/IoTDBTableMetaModel.java:
```diff
@@ -171,26 +175,28 @@ private String getTableDDLInfoWithISPrivilege(JDBCSession session,
+                    toAppend.append("\n) COMMENT '").append(tableComment).append("' ");
+                    toAppend.append("WITH (TTL=").append(ttl).append(");");
+                } else {
+                    toAppend.append("\n) WITH (TTL=").append(ttl).append(");");
```
Append `\n)` unconditionally before this `if` block so this TTL clause can be moved after the block.

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/model/meta/IoTDBTableMetaModel.java:
```diff
@@ -203,29 +209,30 @@ private String getTableDDLInfoWithoutISPrivilege(JDBCSession session,
-            String columnComment = rs.getString("Comment");
-            if (columnComment != null && !columnComment.isEmpty()) {
-                toAppend.append(" COMMENT '").append(columnComment).append("'");
+        try (JDBCStatement stmt = session.createStatement()) {
```
Same as the previous comment. It also looks vaguely similar, can shared parts be extracted and reused?

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/model/IoTDBAbstractUser.java:
```diff
@@ -112,29 +112,31 @@ public void loadGrants(DBRProgressMonitor monitor) throws DBException {
+
+                        String privilegeName = rs.getString("Privileges");
+                        IoTDBPrivilege pri = dataSource.getPrivilege(monitor, privilegeName);
+                        privileges.add(pri);
```
Since it has just a single value in the list, consider using `List#of`

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/model/IoTDBDataSource.java:
```diff
@@ -85,22 +85,26 @@ private List<IoTDBAbstractUser> loadUsers(DBRProgressMonitor monitor) throws DBE
-                    break;
+            try (JDBCStatement stmt = session.createStatement()) {
+                try (JDBCResultSet rs = stmt.executeQuery(sql)) {
+                    while (rs.next()) {
```
This likely should be `if`, not `while`.

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/model/IoTDBDataSource.java:
```diff
@@ -85,22 +85,26 @@ private List<IoTDBAbstractUser> loadUsers(DBRProgressMonitor monitor) throws DBE
+                    try (JDBCStatement stmt2 = session.createStatement()) {
+                        try (JDBCResultSet rs2 = stmt2.executeQuery(sql)) {
+                            while (rs2.next()) {
+                                if (rs2.getString("Privileges").equals("MANAGE_USER")) {
```
In case the result of `rs2.getString("Privileges")` is `null`, this will throw an NPE. Consider swapping operands

```suggestion
                                if ("MANAGE_USER".equals(rs2.getString("Privileges"))) {
```

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/model/IoTDBDataSource.java:
```diff
@@ -114,18 +118,20 @@ private List<IoTDBAbstractUser> loadUsers(DBRProgressMonitor monitor) throws DBE
+                try (JDBCResultSet rs = stmt.executeQuery(sql)) {
+                    while (rs.next()) {
+                        String tmpUserName = rs.getString("User");
+                        if (tmpUserName.equals(currentUserName)) {
```
Same as in the previous comment

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/model/IoTDBRelationalUser.java:
```diff
@@ -78,27 +78,31 @@ public void loadDatabases(DBRProgressMonitor monitor) throws DBException {
+
+                        sql = isTree ? ("show devices " + currentDatabase + ".**") : ("show tables in " + currentDatabase);
+                        try (JDBCStatement stmt2 = session.createStatement()) {
+                            try (JDBCResultSet rs2 = stmt2.executeQuery(sql)) {
```
Is it possible to accumulate a list of devices and query them all at once instead of querying a single device at a time?

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/IoTDBDataSourceProvider.java:
```diff
@@ -83,21 +83,30 @@ private String buildUrlFromTemplate(DBPConnectionConfiguration connectionInfo, S
         return url.toString();
     }
 
+    private String processUrl(String url) {
```
```suggestion
    @NotNull
    private String processUrl(@NotNull String url) {
```

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/IoTDBDataSourceProvider.java:
```diff
@@ -83,21 +83,30 @@ private String buildUrlFromTemplate(DBPConnectionConfiguration connectionInfo, S
+        String processedUrl = url;
+        int index = url.indexOf("?");
+        if (index > 0 && url.charAt(index - 1) == '/') {
+            processedUrl = url.substring(0, index - 1).concat(url.substring(index));
```
Redundant local variable
```suggestion
            return url.substring(0, index - 1).concat(url.substring(index));
```

ShadelessFox commented:
> Hi, this PR fixes an issue in the IoTDB plugin. Could you please help review it and let me know if any changes are needed? Thank you!

Hi, sure! Sorry for the long wait. Unfortunately, your PR went unnoticed. Thanks for letting me know!

ShadelessFox reviewed (commented):
I understand that this is mostly the original code, but since you're already working on it, we can further improve it. :^)

LJW21-02 committed:
Fix review suggestions

LJW21-02 commented:
Hi @ShadelessFox ! Thank you for your review. I have addressed all the suggestions mentioned above, please let me know if anything else needs adjustment!

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/model/IoTDBRelationalUser.java:
```diff
@@ -74,37 +80,67 @@ public List<IoTDBDatabase> getDatabases() {
+                            List<String> currentTables = new ArrayList<>();
+                            currentTables.add(currentTableName);
+                            databaseDevicesMap.put(currentDatabase, currentTables);
+                        }
```
You could use `Map#computeIfAbsent` here
```suggestion
                        databaseDevicesMap
                            .computeIfAbsent(currentDatabase, ignored -> new ArrayList<>())
                            .add(currentTableName);
```

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/model/meta/IoTDBMetaModel.java:
```diff
@@ -59,21 +59,25 @@ public String getTableDDL(@NotNull DBRProgressMonitor monitor,
-                isAligned = (rs.getString("IsAligned")).equals("true");
+            try (JDBCStatement stmt = session.createStatement()) {
+                try (JDBCResultSet rs = stmt.executeQuery(sql)) {
+                    if (rs.next()) {
```
`executeQuery` might return `null`. Pretty sure it's not the case here, but it's better to be sure.
```suggestion
                    if (rs != null && rs.next()) {
```

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/model/meta/IoTDBMetaModel.java:
```diff
@@ -59,21 +59,25 @@ public String getTableDDL(@NotNull DBRProgressMonitor monitor,
-                ddl.append("delete timeseries ").append(rs.getString("Timeseries")).append(";\n");
+            try (JDBCStatement stmt = session.createStatement()) {
+                try (JDBCResultSet rs = stmt.executeQuery(sql)) {
+                    while (rs.next()) {
```
`executeQuery` might return `null`. Pretty sure it's not the case here, but it's better to be sure.
```suggestion
                    while (rs != null && rs.next()) {
```

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/model/meta/IoTDBMetaModel.java:
```diff
@@ -59,21 +59,25 @@ public String getTableDDL(@NotNull DBRProgressMonitor monitor,
+            try (JDBCStatement stmt = session.createStatement()) {
+                try (JDBCResultSet rs = stmt.executeQuery(sql)) {
+                    while (rs.next()) {
+                        ddl.append("delete timeseries ").append(rs.getString("Timeseries")).append(";\n");
```
Consider quoting `rs.getString("Timeseries")` using `DBUtils#getQuotedIdentifier`

LJW21-02 replied:
Thank you for the suggestion.
In IoTDB-tree, the value returned by rs.getString("Timeseries") represents a path expression rather than a relational identifier. Quoting it would change the semantics of the SQL statement and lead to incorrect parsing in tree model queries.
Therefore, DBUtils#getQuotedIdentifier is only applied in the IoTDB-table model, where identifiers follow relational semantics and quoting is required.

ShadelessFox replied:
Thanks for the explanation. Does it also mean that turning that expression into a string will also not work? If it does work, you could use `SQLUtils#quoteString`. The same applies to other occurrences.

LJW21-02 replied:
Thanks for your suggestion, but string conversion won't work here either, so I'll maintain the current implementation.

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/model/meta/IoTDBMetaModel.java:
```diff
@@ -59,21 +59,25 @@ public String getTableDDL(@NotNull DBRProgressMonitor monitor,
         }
 
         try (JDBCSession session = DBUtils.openMetaSession(monitor, (DBSObject) sourceObject, "Execute SQL for IoTDB-tree")) {
             String sql = String.format("show timeseries %s.**", device);
```
Consider quoting `device` using `DBUtils#getQuotedIdentifier`

LJW21-02 replied:
This and the remaining occurrences follow the same reasoning explained above, so they are left unchanged.

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/model/meta/IoTDBMetaModel.java:
```diff
@@ -83,25 +87,27 @@ public String getTableDDL(@NotNull DBRProgressMonitor monitor,
 
         try (JDBCSession session = DBUtils.openMetaSession(monitor, (DBSObject) sourceObject, "Execute SQL for IoTDB-tree")) {
             String sql = String.format("show timeseries %s.**", device);
```
Consider quoting `device` using `DBUtils#getQuotedIdentifier`

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/model/meta/IoTDBMetaModel.java:
```diff
@@ -83,25 +87,27 @@ public String getTableDDL(@NotNull DBRProgressMonitor monitor,
+                try (JDBCResultSet rs = stmt.executeQuery(sql)) {
+                    if (isAligned) {
+                        String prefix = device + ".";
+                        ddl.append("create aligned timeseries ").append(device).append("(");
```
Consider quoting `device` using `DBUtils#getQuotedIdentifier`

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/model/meta/IoTDBMetaModel.java:
```diff
@@ -83,25 +87,27 @@ public String getTableDDL(@NotNull DBRProgressMonitor monitor,
+                        while (rs.next()) {
+                            ddl.append("create timeseries ").append(rs.getString("Timeseries"));
+                            ddl.append(" with datatype=").append(rs.getString("DataType"));
+                            ddl.append(", encoding=").append(rs.getString("Encoding")).append(";\n");
```
Consider quoting values using `DBUtils#getQuotedIdentifier` for identifiers and `SQLUtils#quoteString` for string literals

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/model/meta/IoTDBMetaModel.java:
```diff
@@ -59,21 +59,25 @@ public String getTableDDL(@NotNull DBRProgressMonitor monitor,
+            try (JDBCStatement stmt = session.createStatement()) {
+                try (JDBCResultSet rs = stmt.executeQuery(sql)) {
+                    if (rs.next()) {
+                        isAligned = (rs.getString("IsAligned")).equals("true");
```
Consider using `JDBCUtils#safeGetBoolean`

ShadelessFox replied:
Why not use `ResultSet#getBoolean`?

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/model/IoTDBDataSource.java:
```diff
@@ -114,18 +118,20 @@ private List<IoTDBAbstractUser> loadUsers(DBRProgressMonitor monitor) throws DBE
-                    continue;
+            try (JDBCStatement stmt = session.createStatement()) {
+                try (JDBCResultSet rs = stmt.executeQuery(sql)) {
+                    while (rs.next()) {
```
```suggestion
                    while (rs != null && rs.next()) {
```

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/model/IoTDBDataSource.java:
```diff
@@ -80,27 +80,31 @@ public List<IoTDBAbstractUser> getUsers(DBRProgressMonitor monitor) throws DBExc
-                    break;
+            try (JDBCStatement stmt = session.createStatement()) {
+                try (JDBCResultSet rs = stmt.executeQuery(sql)) {
+                    if (rs.next()) {
```
```suggestion
                    if (rs != null && rs.next()) {
```

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/model/IoTDBDataSource.java:
```diff
@@ -80,27 +80,31 @@ public List<IoTDBAbstractUser> getUsers(DBRProgressMonitor monitor) throws DBExc
+                    sql = "list privileges of user " + currentUserName;
+                    try (JDBCStatement stmt2 = session.createStatement()) {
+                        try (JDBCResultSet rs2 = stmt2.executeQuery(sql)) {
+                            while (rs2.next()) {
```
```suggestion
                            while (rs2 != null && rs2.next()) {
```

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/model/IoTDBDataSource.java:
```diff
@@ -80,27 +80,31 @@ public List<IoTDBAbstractUser> getUsers(DBRProgressMonitor monitor) throws DBExc
+                        userList.add(user);
+                    }
+
+                    sql = "list privileges of user " + currentUserName;
```
Escape `currentUserName` using `DBUtils#getQuotedIdentifier`

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/model/IoTDBRelationalUser.java:
```diff
@@ -74,37 +80,67 @@ public List<IoTDBDatabase> getDatabases() {
+                        String currentDatabase = rs.getString("Database");
+                        List<String> currentTables = new ArrayList<>();
+
+                        sql = "show tables in " + currentDatabase;
```
Escape `currentDatabase` using `DBUtils#getQuotedIdentifier`

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/model/IoTDBRelationalUser.java:
```diff
@@ -74,37 +80,67 @@ public List<IoTDBDatabase> getDatabases() {
+        try (JDBCSession session = DBUtils.openMetaSession(monitor, this, "Load Databases and Tables Info")) {
+            try (JDBCStatement stmt = session.createStatement()) {
+                try (JDBCResultSet rs = stmt.executeQuery("show devices root.** with database")) {
+                    while (rs.next()) {
```
```suggestion
                    while (rs != null && rs.next()) {
```

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/model/meta/IoTDBMetaModel.java:
```diff
@@ -83,25 +87,27 @@ public String getTableDDL(@NotNull DBRProgressMonitor monitor,
+                            ddl.append("create timeseries ").append(rs.getString("Timeseries"));
+                            ddl.append(" with datatype=").append(rs.getString("DataType"));
+                            ddl.append(", encoding=").append(rs.getString("Encoding")).append(";\n");
+                        }
```
Consider quoting values using `DBUtils#getQuotedIdentifier` for identifiers and `SQLUtils#quoteString` for string literals

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/model/meta/IoTDBTableMetaModel.java:
```diff
@@ -138,18 +140,22 @@ private String getTTL(DBRProgressMonitor monitor,
-                ttl = rs.getString("ttl(ms)");
+            try (JDBCStatement stmt = session.createStatement()) {
+                try (JDBCResultSet rs = stmt.executeQuery(sql)) {
+                    if (rs.next()) {
```
```suggestion
                    if (rs != null && rs.next()) {
```

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/model/meta/IoTDBTableMetaModel.java:
```diff
@@ -138,18 +140,22 @@ private String getTTL(DBRProgressMonitor monitor,
-                    ttl = rs.getString("TTL(ms)");
+                try (JDBCStatement stmt = session.createStatement()) {
+                    try (JDBCResultSet rs = stmt.executeQuery(sql)) {
+                        if (rs.next()) {
```
```suggestion
                        if (rs != null && rs.next()) {
```

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/model/meta/IoTDBTableMetaModel.java:
```diff
@@ -138,18 +140,22 @@ private String getTTL(DBRProgressMonitor monitor,
             }
         } catch (Exception e) {
             try (JDBCSession session = DBUtils.openMetaSession(monitor, (DBSObject) sourceObject, "Get IoTDB table details")) {
                 String sql = String.format("show tables details from %s", databaseName);
```
Escape `databaseName` using `DBUtils#getQuotedIdentifier`

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/model/meta/IoTDBTableMetaModel.java:
```diff
@@ -168,63 +174,53 @@ private String getTableDDLInfoWithISPrivilege(JDBCSession session,
+                    String columnComment = rs.getString(columnTitles.get(3));
+                    if (columnComment != null && !columnComment.isEmpty()) {
+                        toAppend.append(" COMMENT '").append(columnComment).append("'");
+                    }
```
Consider quoting values using `DBUtils#getQuotedIdentifier` for identifiers and `SQLUtils#quoteString` for string literals

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/model/meta/IoTDBMetaModel.java:
```diff
@@ -83,25 +87,27 @@ public String getTableDDL(@NotNull DBRProgressMonitor monitor,
+            try (JDBCStatement stmt = session.createStatement()) {
+                try (JDBCResultSet rs = stmt.executeQuery(sql)) {
+                    if (isAligned) {
+                        String prefix = device + ".";
```
Consider quoting `device` using `DBUtils#getQuotedIdentifier`

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/model/meta/IoTDBTableMetaModel.java:
```diff
@@ -168,63 +174,53 @@ private String getTableDDLInfoWithISPrivilege(JDBCSession session,
+            try (JDBCResultSet rs = stmt.executeQuery(sql)) {
+                toAppend.append("CREATE TABLE ").append(insertTableName).append(" (\n");
+                boolean hasColumn = false;
+                while (rs.next()) {
```
```suggestion
                while (rs != null && rs.next()) {
```

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/model/meta/IoTDBTableMetaModel.java:
```diff
@@ -168,63 +174,53 @@ private String getTableDDLInfoWithISPrivilege(JDBCSession session,
+        StringBuilder toAppend = new StringBuilder(200);
+        try (JDBCStatement stmt = session.createStatement()) {
+            try (JDBCResultSet rs = stmt.executeQuery(sql)) {
+                toAppend.append("CREATE TABLE ").append(insertTableName).append(" (\n");
```
Consider quoting `insertTableName` using `DBUtils#getQuotedIdentifier`

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/model/meta/IoTDBTableMetaModel.java:
```diff
@@ -168,63 +174,53 @@ private String getTableDDLInfoWithISPrivilege(JDBCSession session,
+                if (hasColumn) {
+                    toAppend.setLength(toAppend.length() - 2);
+                }
+                String tableComment = ((DBSEntity) sourceObject).getDescription();
```
Redundant cast
```suggestion
                String tableComment = sourceObject.getDescription();
```

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/model/IoTDBAbstractUser.java:
```diff
@@ -112,29 +112,31 @@ public void loadGrants(DBRProgressMonitor monitor) throws DBException {
+                    List<IoTDBGrant> globalPri = new ArrayList<>();
+                    List<IoTDBGrant> schemaPri = new ArrayList<>();
+
+                    while (rs.next()) {
```
```suggestion
                    while (rs != null && rs.next()) {
```

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/model/IoTDBRelationalUser.java:
```diff
@@ -74,37 +80,67 @@ public List<IoTDBDatabase> getDatabases() {
+                        sql = "show tables in " + currentDatabase;
+                        try (JDBCStatement stmt2 = session.createStatement()) {
+                            try (JDBCResultSet rs2 = stmt2.executeQuery(sql)) {
+                                while (rs2.next()) {
```
```suggestion
                                while (rs2 != null && rs2.next()) {
```

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/model/IoTDBRelationalUser.java:
```diff
@@ -74,37 +80,67 @@ public List<IoTDBDatabase> getDatabases() {
-                        currentTables.add(rs2.getString("TableName"));
+            try (JDBCStatement stmt = session.createStatement()) {
+                try (JDBCResultSet rs = stmt.executeQuery(sql)) {
+                    while (rs.next()) {
```
```suggestion
                    while (rs != null && rs.next()) {
```

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/model/IoTDBRelationalUser.java:
```diff
@@ -74,37 +80,67 @@ public List<IoTDBDatabase> getDatabases() {
+     *
+     * @throws DBException if an error occurs
+     */
+    public void loadTreeDatabasesAndDevices(DBRProgressMonitor monitor) throws DBException {
```
```suggestion
    public void loadTreeDatabasesAndDevices(@NotNull DBRProgressMonitor monitor) throws DBException {
```

LJW21-02 committed:
dbeaver/dbeaver#39997 Fix review suggestions: part 1

LJW21-02 committed:
dbeaver/dbeaver#39997 Fix review suggestions: part 1

LJW21-02 committed:
dbeaver/dbeaver#39997 Fix review suggestions: part 3

LJW21-02 commented:
Hi @ShadelessFox ! Thank you once again for your review. I've addressed all the feedback; some are not implemented —— I've left comments explaining why. Please let me know if you have any further questions or feedback.

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/model/meta/IoTDBTableMetaModel.java:
```diff
@@ -168,63 +174,53 @@ private String getTableDDLInfoWithISPrivilege(JDBCSession session,
+                                      String insertTableName,
+                                      String ttl,
+                                      String sql,
+                                      List<String> columnTitles) throws SQLException {
```
```suggestion
    @NotNull
    private String getTableDDLWithSQL(
        @NotNull JDBCSession session,
        @NotNull GenericTableBase sourceObject,
        @NotNull String insertTableName,
        @NotNull String ttl,
        @NotNull String sql,
        @NotNull List<String> columnTitles
    ) throws SQLException {
```

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.iotdb/src/org/jkiss/dbeaver/ext/iotdb/model/meta/IoTDBMetaModel.java:
```diff
@@ -59,21 +59,27 @@ public String getTableDDL(@NotNull DBRProgressMonitor monitor,
+            try (JDBCStatement stmt = session.createStatement()) {
+                try (JDBCResultSet rs = stmt.executeQuery(sql)) {
+                    if (rs != null && rs.next()) {
+                        isAligned = (rs.getBoolean("IsAligned"));
```
```suggestion
                        isAligned = rs.getBoolean("IsAligned");
```

LJW21-02 replied:
Hi, thanks for catching this. However, I just found out that ResultSet#getBoolean is not applicable for the field "isAligned", so I've reverted to using `ResultSet#getString()` to ensure compatibility and avoid errors.

ShadelessFox reviewed (commented):
Several minor nitpicks, and that's about it. Please see my reply in https://github.com/dbeaver/dbeaver/pull/40000#discussion_r2676705698

LJW21-02 committed:
dbeaver/dbeaver#39997 Fix review suggestions

LJW21-02 commented:
@ShadelessFox Thank you for the thorough review! I've addressed all the latest feedback.

ShadelessFox reviewed (approved):
Thanks for the contribution! The PR will now be passed over to our QA team, but it might take a week or more before it's tested and merged.

MashaKorax committed:
Merge branch 'devel' into fix_20251225

MashaKorax committed:
Merge branch 'devel' into fix_20251225
